NIFI-14900 - Adding support for dropStateKeySupported Stateful flag

pvillard31 opened the pull request:
# Summary

NIFI-14900 - Adding support for dropStateKeySupported Stateful flag

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `./mvnw clean install -P contrib-check`
  - [ ] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

pvillard31 commented:
Understood what was happening... When running locally, things are done really fast so, even though there is a 2 nodes cluster, the state is updated at the same time by the processor on each node leading to state value of 1. In Github testing, the processor instances are likely not executing at the exact same time so each instance on each node increments the state once, leading to a value of 2. Given what we want to test here, I just used the PrimaryNodeOnly annotation to solve the problem and we still have the appropriate testing of the new feature.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessorResource.java:
```diff
@@ -489,7 +489,7 @@ public Response getState(
      */
     @POST
-    @Consumes(MediaType.WILDCARD)
+    @Consumes(MediaType.APPLICATION_JSON)
```
Following the pattern of other Resource methods, it looks like this should support both wildcard and JSON, correct?

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/dao/ControllerServiceDAO.java:
```diff
@@ -150,6 +151,7 @@ public interface ControllerServiceDAO {
+     * @param componentStateDTO   state of the controller service
      */
-    void clearState(String controllerServiceId);
+    void clearState(final String controllerServiceId, final ComponentStateDTO componentStateDTO);
```
Minor, but interface method arguments do not need `final`

exceptionfactory commented on nifi-system-tests/nifi-system-test-extensions-bundle/nifi-system-test-extensions/src/main/java/org/apache/nifi/processors/tests/system/MultiKeyState.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return new HashSet<>(Collections.singletonList(REL_SUCCESS));
```
```suggestion
        return Set.of(REL_SUCCESS);
```

exceptionfactory commented on nifi-system-tests/nifi-system-test-extensions-bundle/nifi-system-test-extensions/src/main/java/org/apache/nifi/processors/tests/system/MultiKeyStateNotDroppable.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return new HashSet<>(Collections.singletonList(REL_SUCCESS));
```
```suggestion
        return Set.of(REL_SUCCESS);
```

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/dao/impl/StandardComponentStateDAO.java:
```diff
@@ -42,38 +46,98 @@ private StateMap getState(final String componentId, final Scope scope) {
         }
     }
 
+    public boolean hasExactlyOneKeyRemoved(Map<String, String> currentState, Map<String, String> newState) {
```
Should this method be `public`?

pvillard31 replied:
It should not, good catch

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/dao/impl/StandardComponentStateDAO.java:
```diff
@@ -42,38 +46,98 @@ private StateMap getState(final String componentId, final Scope scope) {
+                // clear both state's at the same time
+                manager.clear(Scope.CLUSTER);
+                manager.clear(Scope.LOCAL);
+            } else if (manager.getState(Scope.LOCAL) != null && !manager.getState(Scope.LOCAL).toMap().isEmpty() && !stateManagerProvider.isClusterProviderEnabled()) {
```
The conditional in each of these clauses is a bit hard to follow. It seems like nesting branches for `componentStateDTO` not null would support a single check for the existing of local state, and then subsequent checks would be easier to follow.

pvillard31 replied:
Pushed a commit with an alternative that is easier to read. It has intermediary returns though. Happy to go back to a full if/else branches approach that would be a bit simplified as you suggested.

exceptionfactory reviewed (changes requested):
Great work on implementing this new feature support @pvillard31. On closer review, the majority of the changes look ready to go. The core details in `StandardComponentStateDAO` are a bit hard to follow, so if that could be reworked a bit, then this should be ready.

pvillard31 commented:
Pushed a commit to add a null safe guard in the state map providers for ZK and k8s. The problem was introduced with the recent decoupling of ZK and was causing the added system tests of this PR to fail although it was OK before the rebase.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/scheduling/RepositoryContextFactory.java:
```diff
@@ -54,7 +54,13 @@ public RepositoryContextFactory(final ContentRepository contentRepository, final
     public RepositoryContext newProcessContext(final Connectable connectable, final AtomicLong connectionIndex) {
-        final StateManager stateManager = stateManagerProvider.getStateManager(connectable.getIdentifier());
+        final StateManager stateManager;
+        if (connectable instanceof org.apache.nifi.controller.ProcessorNode processorNode) {
```
Is there a reason this is limited to Processors? Other components can also use state, and the annotation evaluation should not need to know about the type of class.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/state/StandardStateManager.java:
```diff
@@ -24,21 +24,34 @@
+        this(localProvider, clusterProvider, componentId, () -> false);
+    }
+
+    public StandardStateManager(final StateProvider localProvider, final StateProvider clusterProvider,
```
Is this constructor needed in addition to the `Supplier` based constructor?

pvillard31 replied:
No longer needed, also a leftover. Removing.

exceptionfactory reviewed (commented):
Thanks for the latest update @pvillard31, the revised approach to tracking support looks to be going in a good direction.

I noted several places where there is a check to determine whether a particular Node is a ProcessorNode, do you think that is needed? For ConnectableTasks, perhaps that is reasonable, but on the other hand, the Stateful annotation evaluation should handle any class, so it seems simpler to avoid the ProcessorNode checks unless it is needed for some other reason.

pvillard31 commented:
Thanks David, I cleaned up the code. Sorry about that, changing the directions a couple of times across so many classes didn't help. I also double checked how we were doing the seeding for other component types and added calls in ExtensionBuilder. While this is not needed for reporting tasks / flow analysis rules, it is needed for controller services because this is the only place we construct its initialization context and wire a StateManager. For consistency I added it on all components to be on the safe side even if it's redundant.
